### PR TITLE
feat: add FillID method for Service, Route and Consumer

### PR DESCRIPTION
**Pull request by czeslavo**
Adds `FillID` methods for Service, Route, and Consumer entities to allow setting deterministic IDs for entities based on their unique properties.

For now, it is only implemented for Service, Route and Consumer as that's what's needed on Kubernetes Ingress Controller side to generate stable IDs for analytics in Konnect.

Part of https://github.com/Kong/kubernetes-ingress-controller/issues/3906.

### CODE REVIEW

File: kong/ids.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package kong
+
+import (
+	"fmt"
+
+	"github.com/gofrs/uuid"
```
Comment (mmorel-35) [inline kong/ids.go:6]: Why using a new dependency,  when `github.com/google/uuid` is already used and offer the possibility to create Uuidv5

Comment (czeslavo) [reply kong/ids.go]: Good point, fixed that.

**Comment by czeslavo** [comment]
Looks like the integration tests are blocked by `TestFillPluginDefaults` failing consistently (not related to the code this PR introduces), https://github.com/Kong/go-kong/pull/316 should fix it. Apart from that, it's ready for review.

File: kong/ids.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	case *Route:
+		return fillIDForRoute(e)
+	case *Consumer:
+		return fillIDForConsumer(e)
+	default:
+		return fmt.Errorf("unsupported entity: '%T'", entity)
+	}
+}
+
+func fillIDForService(s *Service) error {
```
Comment (mmorel-35) [inline kong/ids.go:34]: I see something else now. Would you consider making FillID an Entity method. For example :

```suggestion
func (s *Service) FillÎD() error {
```

File: kong/ids.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	"fmt"
+	"reflect"
+
+	"github.com/google/uuid"
+)
+
+// IDFillableEntitiesT is a type constraint for entities that can be filled with an ID.
+type IDFillableEntitiesT interface {
+	Service | Route | Consumer
+}
```
Comment (mmorel-35) [inline kong/ids.go:13]:
```suggestion
// IDFillable is a type constraint for entities that can be filled with an ID.
type IDFillable interface {
	FillÎD() error
}
```

File: kong/ids.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	gen, err := idGeneratorFor(*s)
+	if err != nil {
+		return fmt.Errorf("could not get id generator: %w", err)
+	}
+
+	s.ID = gen(*s.Name)
+	return nil
+}
+
+func fillIDForRoute(r *Route) error {
```
Comment (mmorel-35) [inline kong/ids.go:55]:
```suggestion
func (r *Route) FillID error {
```

File: kong/ids.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	gen, err := idGeneratorFor(*r)
+	if err != nil {
+		return fmt.Errorf("could not get id generator: %w", err)
+	}
+
+	r.ID = gen(*r.Name)
+	return nil
+}
+
+func fillIDForConsumer(c *Consumer) error {
```
Comment (mmorel-35) [inline kong/ids.go:76]:
```suggestion
func (c *Consumer) FillID error {
```

File: kong/ids_test.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	name   string
+	entity *T
+
+	assertEntity func(t *testing.T, entity *T)
+	expectErr    bool
+}
+
+func (tc fillEntityIDTestCase[T]) run(t *testing.T) {
+	t.Run(tc.name, func(t *testing.T) {
+		err := kong.FillEntityID(tc.entity)
```
Comment (mmorel-35) [inline kong/ids_test.go:21]:
```suggestion
		err := tc.entity.FillID()
```

File: kong/ids_test.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/kong/go-kong/kong"
+)
+
+type fillEntityIDTestCase[T kong.IDFillableEntitiesT] struct {
+	name   string
+	entity *T
```
Comment (mmorel-35) [inline kong/ids_test.go:13]:
```suggestion
	entity *IDFillable
```

File: kong/ids.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+type idGenerator = func(string) *string
+
+// newIDGeneratorFor returns a new ID generator for the given entity type. Should be used only to initialize _idGenerators.
+func newIDGeneratorFor(entityPluralName string) idGenerator {
+	entityTypeNamespace := uuid.NewSHA1(_kongEntitiesNamespace, []byte(entityPluralName))
+	return func(entityKey string) *string {
+		id := uuid.NewSHA1(entityTypeNamespace, []byte(entityKey)).String()
+		return &id
+	}
+}
```
Comment (mmorel-35) [inline kong/ids.go:121]:
```suggestion
type idGenerator struct {
	Namespace uuid.UUID
}

func (i idGenerator) BuildIDFor(entityKey string) *string {
	return &uuid.NewSHA1(i.Namespace, []byte(entityKey)).String()
}

// newIDGeneratorFor returns a new ID generator for the given entity type. Should be used only to initialize _idGenerators.
func newIDGeneratorFor(entityPluralName string) idGenerator {
	return idGenerator{Namespace: uuid.NewSHA1(_kongEntitiesNamespace, []byte(entityPluralName))}
}
```

File: kong/ids.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+	if c.Username == nil || *c.Username == "" {
+		return fmt.Errorf("consumer username is required")
+	}
+
+	gen, err := idGeneratorFor(*c)
+	if err != nil {
+		return fmt.Errorf("could not get id generator: %w", err)
+	}
+
+	c.ID = gen(*c.Username)
```
Comment (mmorel-35) [inline kong/ids.go:93]:
```suggestion
	c.ID = gen.BuildIDFor(*c.Username)
```

File: kong/ids.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+	if r.Name == nil || *r.Name == "" {
+		return fmt.Errorf("route name is required")
+	}
+
+	gen, err := idGeneratorFor(*r)
+	if err != nil {
+		return fmt.Errorf("could not get id generator: %w", err)
+	}
+
+	r.ID = gen(*r.Name)
```
Comment (mmorel-35) [inline kong/ids.go:72]:
```suggestion
	r.ID = gen.BuildIDFor(*r.Name)
```

File: kong/ids.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	if s.Name == nil || *s.Name == "" {
+		return fmt.Errorf("service name is required")
+	}
+
+	gen, err := idGeneratorFor(*s)
+	if err != nil {
+		return fmt.Errorf("could not get id generator: %w", err)
+	}
+
+	s.ID = gen(*s.Name)
```
Comment (mmorel-35) [inline kong/ids.go:51]:
```suggestion
	s.ID = gen.BuildIDFor(*s.Name)
```

File: kong/ids.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+	case *Service:
+		return fillIDForService(e)
+	case *Route:
+		return fillIDForRoute(e)
+	case *Consumer:
+		return fillIDForConsumer(e)
+	default:
+		return fmt.Errorf("unsupported entity: '%T'", entity)
+	}
+}
```
Comment (mmorel-35) [inline kong/ids.go:32]: WIth the use of IDFillable interface this should not be of use anymore

```suggestion
```

File: kong/ids.go (hunk 2 above)
Comment (czeslavo) [reply kong/ids.go:34]: Yeah, that will also work. I was just trying to mimic the pattern that was used for `FillEntityDefaults` and `FillPluginsDefaults`.

**Comment by czeslavo** [comment]
@mmorel-35 I hope I addressed all your suggestions, PTAL.

**Review by mmorel-35 — APPROVED** [review]

**Comment by mmorel-35** [comment]
Would also need to do it for plugins ?
In that case it the `buildIDFor()` could take as parameter the concatenation of the plugin name with the the ID of the referenced entity (Consumer, Route or Service) or just the name of the plugin when it's a global one.
I would personally add a character separator between the referenced ID and the plugin name but it might not be absolutely necessary

File: kong/ids_test.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/kong/go-kong/kong"
+)
+
+type idFillable interface {
+	FillID() error
+}
```
Comment (mmorel-35) [inline kong/ids_test.go:13]: Oh I missed this one !
Can't you make idFillable public in kong/ids.go ? 
This way you can use it here with kong.IDFillable

Comment (czeslavo) [reply kong/ids_test.go:13]: Yeah, my point was to just reduce public API to a minimum, but I guess exporting an interface won't hurt us.

Comment (czeslavo) [reply kong/ids_test.go]: https://github.com/Kong/go-kong/pull/299/commits/ce04fc3eb240e1abaf3f065592feac6887d36917

**Review by mmorel-35 — APPROVED** [review]

**Review by pmalek — APPROVED** [review]

**Pull request merged**